### PR TITLE
Ports IPCs from beestation

### PR DESCRIPTION
**Pull request by redmoogle**
# Document the changes in your pull request

# Seems to work and be bug free

Credits to BeeStation(Credits would be a milelong if I listed each individual port)

Adds IPCs
They're a synthetic race **different** from preternis, mainly with chassis customization and TV screens. They have a high burn mod
also they cant be cloned because they're robots. so you can use a cyborg rebooter in place of that.
They lose limbs easily but can be attached easily, Head is stored in the _chest_ and _can_ live without a head

Wiki Documentation:

https://wiki.yogstation.net/wiki/User:Redmoogle

# Spriting
Various types of Antennas and Displays

image
image
image
image
image
image
image
image
image
image

# Changelog

:cl:  Redmoogle, BeeStation, Shiptest
rscadd: Add IPCs from Beestation
/:cl:

### CODE REVIEW

**Comment by ynot01** [comment]
dejavu

**Comment by Moggram** [comment]
We have been over this before and it was shot down.

**Comment by redmoogle** [comment]
> We have been over this before and it was shot down.

It wasnt shot down. I just stopped working on it

**Comment by Spacemanspark** [comment]
@Moggram for bad code and bugs from what I've heard, not for the feature itself.

Now, I'm curious; what's the extent of bee IPC customization?

And I might need a bigger breakdown of how bee IPCs work. If they really explode on death or some of the other things I've heard... ehhhh.

File: code/modules/mob/living/carbon/human/species_types/IPC.dm [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	exotic_blood = "oil"
+	damage_overlay_type = "synth"
+	limbs_id = "synth"
+	burnmod = 2
+	heatmod = 1.5
+	brutemod = 1
+	toxmod = 0
+	clonemod = 0
+	staminamod = 0.8
+	siemens_coeff = 1.5
```
Comment (Anvilman6) [inline code/modules/mob/living/carbon/human/species_types/IPC.dm]:
```suggestion
	siemens_coeff = 1.75 //Matches Preternis
	payday_modifier = 0.6 //Clunky robots are valued less
```

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 2]
```diff
@@ -433,6 +433,39 @@
...
+	description = "Repairs brain damage in synthetics."
+	color = "#727272"
+	taste_description = "metallic"
+	process_flags = SYNTHETIC
+
+/datum/reagent/medicine/liquid_solder/on_mob_life(mob/living/M)
+	M.adjustOrganLoss(ORGAN_SLOT_BRAIN, -3*REM)
+	if(iscarbon(M))
+		var/mob/living/carbon/C = M
+		if(prob(30) && C.has_trauma_type(BRAIN_TRAUMA_SPECIAL))
```
Comment (Anvilman6) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: Higher chance to fix special traumas over minor?

File: code/modules/reagents/chemistry/recipes/medicine.dm [hunk 3]
```diff
@@ -276,6 +276,20 @@
 	results = list(/datum/reagent/medicine/psicodine = 5)
 	required_reagents = list( /datum/reagent/medicine/mannitol = 2, /datum/reagent/water = 2, /datum/reagent/impedrezene = 1)
 
+/datum/chemical_reaction/system_cleaner
+	name = "System Cleaner"
+	id = /datum/reagent/medicine/system_cleaner
+	results = list(/datum/reagent/medicine/system_cleaner = 4)
```
Comment (Anvilman6) [inline code/modules/reagents/chemistry/recipes/medicine.dm]: 5 components : 4 products is kind of an odd ratio

**Review by Anvilman6 — COMMENTED** [review]
Quick partial review

**Comment by Spacemanspark** [comment]
Glory to Synthetica

**Comment by AlvCyktor** [comment]
ayo???? AYO?????

**Comment by Hislyna** [comment]
I love this and I really hope it comes in the game~ I did notice on a private server that the special organs that IPCs have such as the cell don't have sprites when I take them out of the body however.

**Comment by redmoogle** [comment]
> I love this and I really hope it comes in the game~ I did notice on a private server that the special organs that IPCs have such as the cell don't have sprites when I take them out of the body however.

Fixed it

**Comment by Hislyna** [comment]
Are the brain surgeries supposed to be on the head, or moved to the chest? Brain Surgery, Brainwashing, Lobotomy and Pacification are still surgeries when aimed at the head

**Comment by Spacemanspark** [comment]
Should any of those even work on IPCs given that they're robots and not imbued with inferior flesh brains

**Comment by Hislyna** [comment]
> Should any of those even work on IPCs given that they're robots and not imbued with inferior flesh brains

That is true, Solder is supposed to be the go to for all the brain shiz, so I guess they shouldnt even be there. Idk what people would think of not being able to brainwash/pacify them. Maybe make it use an Law upload board for a step in brainwashing?

**Comment by redmoogle** [comment]
> Are the brain surgeries supposed to be on the head, or moved to the chest? Brain Surgery, Brainwashing, Lobotomy and Pacification are still surgeries when aimed at the head

I can always make some additions to surgery to make it be able to do certain surgery for other than the *normal* slots

**Comment by Hislyna** [comment]
> > Are the brain surgeries supposed to be on the head, or moved to the chest? Brain Surgery, Brainwashing, Lobotomy and Pacification are still surgeries when aimed at the head
> 
> I can always make some additions to surgery to make it be able to do certain surgery for other than the _normal_ slots

Up to you at the end of the day, either way its looking good~

**Comment by Kapu1178** [comment]
oh you poor soul.

This code is extremely cursed and really not worth porting. It has so many issues and hacky work arounds to this day, and introduces alot of tech debt. I am responsible for fixing up a large part of it but I'd still recommend just remaking them from scratch, it's unsalvagable.

**Comment by redmoogle** [comment]
> oh you poor soul.
> 
> This code is extremely cursed and really not worth porting. It has so many issues and hacky work arounds to this day, and introduces alot of tech debt. I am responsible for fixing up a large part of it but I'd still recommend just remaking them from scratch, it's unsalvagable.

that would likely rewrite a large amount of organ/surgery code which I would rather do after this

**Comment by Kapu1178** [comment]
> > oh you poor soul.
> > This code is extremely cursed and really not worth porting. It has so many issues and hacky work arounds to this day, and introduces alot of tech debt. I am responsible for fixing up a large part of it but I'd still recommend just remaking them from scratch, it's unsalvagable.
> 
> that would likely rewrite a large amount of organ/surgery code which I would rather do after this

as the guy who *did* that, I really, really suggest against it. I also took a peek at the code and it looks like you're using an outdated version which lacks my optimizations and bugfixes. Especially the one to reagent processing.

File: code/modules/mob/living/carbon/human/species_types/IPC.dm [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	mutantstomach = /obj/item/organ/stomach/cell
+	mutantears = /obj/item/organ/ears/robot
+	mutant_organs = list(/obj/item/organ/cyberimp/arm/power_cord)
+	mutant_bodyparts = list("ipc_screen", "ipc_antenna", "ipc_chassis")
+	default_features = list("mcolor" = "#7D7D7D", "ipc_screen" = "Static", "ipc_antenna" = "None", "ipc_chassis" = "Morpheus Cyberkinetics(Greyscale)")
+	meat = /obj/item/stack/sheet/plasteel{amount = 5}
+	skinned_type = /obj/item/stack/sheet/metal{amount = 10}
+	exotic_blood = /datum/reagent/oil
+	damage_overlay_type = "synth"
+	limbs_id = "synth"
```
Comment (Anvilman6) [inline code/modules/mob/living/carbon/human/species_types/IPC.dm:23]:
```suggestion
	limbs_id = "synth"
	payday_modifier = 0.6 //Mass producible labor
```

Comment (Anvilman6) [reply code/modules/mob/living/carbon/human/species_types/IPC.dm:23]: Adds payday modifier for consistency with other species

**Review by Anvilman6 — CHANGES_REQUESTED** [review]
Payday Modifier

File: code/modules/mob/living/carbon/human/species_types/IPC.dm (hunk 4 above)
Comment (Anvilman6) [reply code/modules/mob/living/carbon/human/species_types/IPC.dm:23]: I'll do a full look through at some point soon

**Comment by MegaEmpirical** [comment]
If you cut their head off and examine, it's as if you're examining their body and it'll say head missing and everything they're wearing.

EDIT: What do you mean the head is in the chest? It is clearly above the torso.

**Comment by Spacemanspark** [comment]
@MegaEmpirical

The _brain_ is in the torso. IPCs can survive without a head, as the positronic is seated in the torso.

**Comment by MegaEmpirical** [comment]
> @MegaEmpirical
> 
> The _brain_ is in the torso. IPCs can survive without a head, as the positronic is seated in the torso.

So they're cyborgs not robots?

**Comment by redmoogle** [comment]
> > @MegaEmpirical
> > The _brain_ is in the torso. IPCs can survive without a head, as the positronic is seated in the torso.
> 
> So they're cyborgs not robots?

what dictates robots not having brains in their torso

**Comment by MegaEmpirical** [comment]
> > > @MegaEmpirical
> > > The _brain_ is in the torso. IPCs can survive without a head, as the positronic is seated in the torso.
> > 
> > 
> > So they're cyborgs not robots?
> 
> what dictates robots not having brains in their torso

Biology. Language.

**Comment by redmoogle** [comment]
> > > > @MegaEmpirical
> > > > The _brain_ is in the torso. IPCs can survive without a head, as the positronic is seated in the torso.
> > > 
> > > 
> > > So they're cyborgs not robots?
> > 
> > 
> > what dictates robots not having brains in their torso
> 
> Biology. Language.

A cyborg would be a mix of humans and robotics (preternis) IPCs are exclusively robotic and so none of that applies

**Comment by MegaEmpirical** [comment]
> > > > > @MegaEmpirical
> > > > > The _brain_ is in the torso. IPCs can survive without a head, as the positronic is seated in the torso.
> > > > 
> > > > 
> > > > So they're cyborgs not robots?
> > > 
> > > 
> > > what dictates robots not having brains in their torso
> > 
> > 
> > Biology. Language.
> 
> A cyborg would be a mix of humans and robotics (preternis) IPCs are exclusively robotic and so none of that applies

Thank you, that's what I needed to know.

**Comment by adamsong** [comment]
Emotes work, but say they don't

**Comment by redmoogle** [comment]
> Emotes work, but say they don't

What

**Comment by adamsong** [comment]
image
But the emote works
Also, defibs work

**Comment by MegaEmpirical** [comment]
How do IPCs interact with zombies?

**Comment by redmoogle** [comment]
> How do IPCs interact with zombies?

they die

**Comment by ynot01** [comment]
IPCs do not have their name properly set when roundstart nukie, and when brought in as a spare operative, are spawned in as a naked IPC without an antag datum or any equipment

**Review by ynot01 — CHANGES_REQUESTED** [review]
IPCs do not have their name properly set when roundstart nukie, and when brought in as a spare operative, are spawned in as a naked IPC without an antag datum or any equipment

**Comment by redmoogle** [comment]
> IPCs do not have their name properly set when roundstart nukie, and when brought in as a spare operative, are spawned in as a naked IPC without an antag datum or any equipment

I dont think IPCs should be spawning as nukies

**Comment by ynot01** [comment]
IPC bloodsuckers when dead will spawn a power cord implant on their turf every tick and can't revive

**Comment by ynot01** [comment]
> I dont think IPCs should be spawning as nukies

yeah probably not but they are

**Comment by redmoogle** [comment]
> IPC bloodsuckers when dead will spawn a power cord implant on their turf every tick and can't revive

that **should not** happen because that means its regenerating organs every tick (as in qdeling and respawning them)

**Comment by Spacemanspark** [comment]
Should IPCs even be spawning as bloodsuckers to begin with?

**Comment by redmoogle** [comment]
> Should IPCs even be spawning as bloodsuckers to begin with?

well we have no system in place to really blacklist a species other than forcing them to another

**Comment by MegaEmpirical** [comment]
I'd support the implementation of an antag racial blacklist or of forcing human when such an antag is rolled.

Similarly, plasmamen should not be vampires, or at the very least be unable to drink blood while wearing a helmet and mask.

Can changelings eat IPCs?

**Comment by redmoogle** [comment]
> I'd support the implementation of an antag racial blacklist or of forcing human when such an antag is rolled.
> 
> Similarly, plasmamen should not be vampires, or at the very least be unable to drink blood while wearing a helmet and mask.
> 
> Can changelings eat IPCs?

This PR is already starting to go out of scope so a later PR

and also No

**Review by JamieD1 — APPROVED** [review]

**Comment by Spacemanspark** [comment]
The glorious Synthetica regime will spread across all of SS13

**Review by ReddicusDragon — COMMENTED** [review]
pog

**Pull request merged**